llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp

```diff
@@ -6449,30 +6450,74 @@ void SelectionDAGBuilder::visitVectorExtractLastActive(const CallInst &I,
   setValue(&I, Result);
 }
 
+static SDValue createMask(SelectionDAG &DAG, const SDLoc &DL, SDValue Cond,
+                          EVT WorkingVT, const SDNodeFlags &Flag) {
+  if (WorkingVT.isVector())
+    return DAG.getSExtOrTrunc(Cond, DL, WorkingVT);
+
+  // Extend cond to WorkingVT and normalize to 0 or 1
+  if (Cond.getValueType() != WorkingVT)
+    Cond = DAG.getNode(ISD::ZERO_EXTEND, DL, WorkingVT, Cond, Flag);
+
+  // Normalize
+  SDValue One = DAG.getConstant(1, DL, WorkingVT);
+  SDValue Norm = DAG.getNode(ISD::AND, DL, WorkingVT, Cond, One, Flag);
+
+  // Mask = 0 - Norm
+  SDValue Zero = DAG.getConstant(0, DL, WorkingVT);
+  return DAG.getNode(ISD::SUB, DL, WorkingVT, Zero, Norm, Flag);
```

Comment:
This may not necessarily always work, but have you checked if you can sign-extend the `i1` cond value, instead of subtracting it from 0 to generate the mask? If it can't be done a comment on the reason would be great, so we know the possible gotchas!